### PR TITLE
Periodic maintenance

### LINKED ISSUE #923: add automatic "maintenance" to stellar-core
stellar-core has a `maintenance` endpoint that should be periodically called to clean up the database.

`maintenance` right now only performs historical pruning of data that was successfully uploaded to history and/or imported into dependent services such as Horizon.

At a minimum, we should call it on startup (probably add a config to control this).

Maybe consider adding an automatic schedule within core (less important).

This can only be implemented after all dependencies such as Horizon properly update their cursor.

### PR DESCRIPTION
**Pull request by vogel**
Fixes #923

### CODE REVIEW

File: src/main/CommandHandler.cpp [hunk 1]
```diff
@@ -969,7 +970,12 @@ CommandHandler::maintenance(std::string const& params, std::string& retStr)
     http::server::server::parseParams(params, map);
     if (map["queue"] == "true")
     {
-        mApp.maintenance(50000);
+        auto count = 50000;
```
Comment (MonsieurNicolas) [inline src/main/CommandHandler.cpp]: you should use `uint32` instead of auto, otherwise you're allowing people to pass in negative numbers

Comment (MonsieurNicolas) [reply src/main/CommandHandler.cpp]: talking about that, it looks like `parseNumParam` has a bug: it doesn't check if the stream failed when parsing the input `str.fail()`

File: src/main/Maintainer.cpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
```
Comment (MonsieurNicolas) [inline src/main/Maintainer.cpp]: new files -> needs to be added to the Visual Studio project files

File: src/main/Config.cpp [hunk 3]
```diff
@@ -321,14 +322,37 @@ Config::load(std::string const& filename)
                 }
                 ALLOW_LOCALHOST_FOR_TESTING = item.second->as<bool>()->value();
             }
-            else if (item.first == "MAINTENANCE_ON_STARTUP")
+            else if (item.first == "AUTOMATIC_MAINTENANCE_PERIOD")
```
Comment (MonsieurNicolas) [inline src/main/Config.cpp]: you need to update `stellar-core_example.cfg`

Comment (MonsieurNicolas) [reply src/main/Config.cpp]: in the comments in the config, include also which values to use to disable it

File: src/main/Config.cpp [hunk 4]
```diff
@@ -321,14 +322,37 @@ Config::load(std::string const& filename)
...
             {
-                if (!item.second->as<bool>())
+                if (!item.second->as<int64_t>())
+                {
+                    throw std::invalid_argument(
+                        "invalid AUTOMATIC_MAINTENANCE_PERIOD");
+                }
+                int64_t parsedAutomaticMaintenancePeriod =
+                    item.second->as<int64_t>()->value();
+                if (parsedAutomaticMaintenancePeriod <= 0 ||
```
Comment (MonsieurNicolas) [inline src/main/Config.cpp]: maybe allow 0 to disable

File: src/main/Maintainer.cpp [hunk 5]
```diff
@@ -3,16 +3,37 @@
...
 
-Maintainer::Maintainer(Application& app) : mApp{app}
+Maintainer::Maintainer(Application& app) : mApp{app}, mTimer{mApp}
 {
 }
 
+void
+Maintainer::start()
+{
+    scheduleMaintenance();
```
Comment (MonsieurNicolas) [inline src/main/Maintainer.cpp]: we could put a condition that `AUTOMATIC_MAINTENANCE_PERIOD > 0` to disable this

File: src/main/Maintainer.cpp [hunk 6]
```diff
@@ -3,16 +3,37 @@
...
+}
+
+void
+Maintainer::scheduleMaintenance()
+{
+    mTimer.expires_from_now(mApp.getConfig().AUTOMATIC_MAINTENANCE_PERIOD);
+    mTimer.async_wait([this]() { tick(); }, VirtualTimer::onFailureNoop);
```
Comment (MonsieurNicolas) [inline src/main/Maintainer.cpp]: I think this is safe (app being destructed vs timer events)

File: src/main/Config.cpp (hunk 4 above)
Comment (MonsieurNicolas) [reply src/main/Config.cpp]: I just realized that I actually specified 0 as the way to disable it in the issue

**Comment by vogel** [comment]
Updated

File: src/main/Config.cpp [hunk 7]
```diff
@@ -321,14 +322,37 @@ Config::load(std::string const& filename)
...
+                        "invalid AUTOMATIC_MAINTENANCE_COUNT");
                 }
-                MAINTENANCE_ON_STARTUP = item.second->as<bool>()->value();
+                int64_t parsedAutomaticMaintenanceCount =
+                    item.second->as<int64_t>()->value();
+                if (parsedAutomaticMaintenanceCount < 0 ||
+                    parsedAutomaticMaintenanceCount > UINT32_MAX)
+                    throw std::invalid_argument(
+                        "invalid AUTOMATIC_MAINTENANCE_COUNT");
+                AUTOMATIC_MAINTENANCE_COUNT = static_cast<unsigned short>(
```
Comment (MonsieurNicolas) [inline src/main/Config.cpp]: cast should be `uint32_t` not `unsigned short`

File: src/main/Config.cpp [hunk 8]
```diff
@@ -321,14 +322,37 @@ Config::load(std::string const& filename)
...
-                if (!item.second->as<bool>())
+                if (!item.second->as<int64_t>())
+                {
+                    throw std::invalid_argument(
+                        "invalid AUTOMATIC_MAINTENANCE_PERIOD");
+                }
+                int64_t parsedAutomaticMaintenancePeriod =
+                    item.second->as<int64_t>()->value();
+                if (parsedAutomaticMaintenancePeriod < 0 ||
+                    parsedAutomaticMaintenancePeriod > UINT32_MAX)
```
Comment (MonsieurNicolas) [inline src/main/Config.cpp]: style: use braces around the `throw` (especially as the if is already spanning multiple lines)

File: src/main/Config.cpp [hunk 9]
```diff
@@ -321,14 +322,37 @@ Config::load(std::string const& filename)
...
                 {
                     throw std::invalid_argument(
-                        "invalid MAINTENANCE_ON_STARTUP");
+                        "invalid AUTOMATIC_MAINTENANCE_COUNT");
                 }
-                MAINTENANCE_ON_STARTUP = item.second->as<bool>()->value();
+                int64_t parsedAutomaticMaintenanceCount =
+                    item.second->as<int64_t>()->value();
+                if (parsedAutomaticMaintenanceCount < 0 ||
+                    parsedAutomaticMaintenanceCount > UINT32_MAX)
```
Comment (MonsieurNicolas) [inline src/main/Config.cpp]: same: braces around `throw`

File: src/main/Config.cpp [hunk 10]
```diff
@@ -166,6 +167,49 @@ Config::loadQset(std::shared_ptr<cpptoml::toml_group> group, SCPQuorumSet& qset,
...
+readString(ConfigItem const& item)
+{
+    if (!item.second->as<std::string>())
+    {
+        throw std::invalid_argument(fmt::format("invalid {}", item.first));
+    }
+    return item.second->as<std::string>()->value();
+}
+
+template <typename T>
```
Comment (MonsieurNicolas) [inline src/main/Config.cpp]: nice addition! long overdue

File: src/main/Config.cpp [hunk 11]
```diff
@@ -189,196 +233,73 @@ Config::load(std::string const& filename)
...
-                {
-                    throw std::invalid_argument("invalid FAILURE_SAFETY");
-                }
-                int64_t f = item.second->as<int64_t>()->value();
-                if (f < -1 || f >= INT32_MAX)
-                {
-                    throw std::invalid_argument("invalid FAILURE_SAFETY");
-                }
-                FAILURE_SAFETY = (int32_t)f;
+                FAILURE_SAFETY = readInt<int32_t>(item, 0, INT32_MAX - 1);
```
Comment (MonsieurNicolas) [inline src/main/Config.cpp]: -1 is valid

File: src/main/Config.cpp [hunk 12]
```diff
@@ -414,116 +331,39 @@ Config::load(std::string const& filename)
...
-                    item.second->as<int64_t>()->value();
-                if (parsedTargetPeerConnections <= 0 ||
-                    parsedTargetPeerConnections > UINT16_MAX)
-                {
-                    throw std::invalid_argument(
-                        "invalid TARGET_PEER_CONNECTIONS");
-                }
-                TARGET_PEER_CONNECTIONS =
-                    static_cast<unsigned short>(parsedTargetPeerConnections);
+                TARGET_PEER_CONNECTIONS = readInt<unsigned short>(item);
```
Comment (MonsieurNicolas) [inline src/main/Config.cpp]: 0 is not allowed

File: src/main/Config.cpp [hunk 13]
```diff
@@ -414,116 +331,39 @@ Config::load(std::string const& filename)
...
-                int64_t parsedMaxPeerConnections =
-                    item.second->as<int64_t>()->value();
-                if (parsedMaxPeerConnections <= 0 ||
-                    parsedMaxPeerConnections > UINT16_MAX)
-                {
-                    throw std::invalid_argument("invalid MAX_PEER_CONNECTIONS");
-                }
-                MAX_PEER_CONNECTIONS =
-                    static_cast<unsigned short>(parsedMaxPeerConnections);
+                MAX_PEER_CONNECTIONS = readInt<unsigned short>(item);
```
Comment (MonsieurNicolas) [inline src/main/Config.cpp]: 0 is not allowed

File: src/main/Config.cpp [hunk 14]
```diff
@@ -592,24 +428,11 @@ Config::load(std::string const& filename)
...
             else if (item.first == "MAX_CONCURRENT_SUBPROCESSES")
             {
-                if (!item.second->as<int64_t>())
-                {
-                    throw std::invalid_argument(
-                        "invalid MAX_CONCURRENT_SUBPROCESSES");
-                }
-                MAX_CONCURRENT_SUBPROCESSES =
-                    (size_t)item.second->as<int64_t>()->value();
+                MAX_CONCURRENT_SUBPROCESSES = readInt<size_t>(item);
```
Comment (MonsieurNicolas) [inline src/main/Config.cpp]: I imagine that 0 should not be allowed

**Comment by vogel** [comment]
Updated

**Comment by MonsieurNicolas** [comment]
r+ 864c7963cd22ed33e6029e473bf4228406dfccfb

**Comment by MonsieurNicolas** [comment]
@latobarita: retry

**Pull request merged**